src/enumerable/enumerable_map.jl

```diff
@@ -1,37 +1,86 @@
+# This is the HasEltype() version
+
+struct EnumerableMapHasEltype{T, S, Q<:Function} <: Enumerable
     source::S
     f::Q
 end
 
-Base.iteratorsize(::Type{EnumerableMap{T,S,Q}}) where {T,S,Q} = Base.iteratorsize(S) in (Base.HasLength(), Base.HasShape()) ? Base.HasLength() : Base.iteratorsize(S)
+Base.iteratorsize(::Type{EnumerableMapHasEltype{T,S,Q}}) where {T,S,Q} = Base.iteratorsize(S) in (Base.HasLength(), Base.HasShape()) ? Base.HasLength() : Base.iteratorsize(S)
 
-Base.eltype(iter::EnumerableMap{T,S,Q}) where {T,S,Q} = T
+Base.eltype(iter::Type{EnumerableMapHasEltype{T,S,Q}}) where {T,S,Q} = T
 
-Base.eltype(iter::Type{EnumerableMap{T,S,Q}}) where {T,S,Q} = T
+Base.length(iter::EnumerableMapHasEltype{T,S,Q}) where {T,S,Q} = length(iter.source)
 
-Base.length(iter::EnumerableMap{T,S,Q}) where {T,S,Q} = length(iter.source)
+function Base.start(iter::EnumerableMapHasEltype{T,S,Q}) where {T,S,Q}
+    s = start(iter.source)
+    return s
+end
 
-function map(source::Enumerable, f::Function, f_expr::Expr)
-    TS = eltype(source)
-    T = Base._return_type(f, Tuple{TS,})
-    S = typeof(source)
-    Q = typeof(f)
-    return EnumerableMap{T,S,Q}(source, f)
+function Base.next(iter::EnumerableMapHasEltype{T,S,Q}, s) where {T,S,Q}
+    x = next(iter.source, s)
+    v = x[1]
+    s_new = x[2]
+    v_new = iter.f(v)::T
+    return v_new, s_new
 end
 
-function Base.start(iter::EnumerableMap{T,S,Q}) where {T,S,Q}
+function Base.done(iter::EnumerableMapHasEltype{T,S,Q}, state) where {T,S,Q}
+    return done(iter.source, state)
+end
+
+# This is the EltypeUnknown() version
+
+struct EnumerableMapEltypeUnknown{S, Q<:Function} <: Enumerable
+    source::S
+    f::Q
+end
+
+Base.iteratorsize(::Type{EnumerableMapEltypeUnknown{S,Q}}) where {S,Q} = Base.iteratorsize(S) in (Base.HasLength(), Base.HasShape()) ? Base.HasLength() : Base.iteratorsize(S)
+
+Base.iteratoreltype(::Type{EnumerableMapEltypeUnknown{S,Q}}) where {S,Q} = Base.EltypeUnknown()
+
+Base.length(iter::EnumerableMapEltypeUnknown) = length(iter.source)
+
+function Base.start(iter::EnumerableMapEltypeUnknown)
     s = start(iter.source)
     return s
 end
 
-function Base.next(iter::EnumerableMap{T,S,Q}, s) where {T,S,Q}
+function Base.next(iter::EnumerableMapEltypeUnknown, s)
     x = next(iter.source, s)
     v = x[1]
     s_new = x[2]
-    v_new = iter.f(v)::T
+    v_new = iter.f(v)
     return v_new, s_new
 end
 
-function Base.done(iter::EnumerableMap{T,S,Q}, state) where {T,S,Q}
+function Base.done(iter::EnumerableMapEltypeUnknown, state)
     return done(iter.source, state)
 end
+
+# Implementation of the query operator
+
+function _map(source::Enumerable, f::Function, f_expr::Expr, ::Base.EltypeUnknown)
+    S = typeof(source)
+    Q = typeof(f)
+    println("Unkonwn")
+    return EnumerableMapEltypeUnknown{S,Q}(source, f)
+end
+
+function _map(source::Enumerable, f::Function, f_expr::Expr, ::Base.HasEltype)
+    TS = eltype(source)
+    T = Base._return_type(f, Tuple{TS,})
```

Comment:
This still relying on inference though? I think to really avoid Inference, we should mirror the implementation of [collect](https://github.com/JuliaLang/julia/blob/50413a46a05b4f522e09fac2fb9b6a705cac31d7/base/array.jl#L548) on Generators and how the new [Broadcast](https://github.com/JuliaLang/julia/pull/25377#issuecomment-382602443) will work.

Reply:
> The strategy here is to use the information from inference if it gives us a concrete type, otherwise not use that information. I think that is the same strategy that Broadcast is using now? As far as I understand it, that would be an ok use of inference.

FWIW, I don't think the question is whether the inferred type is concrete or not. It's rather that the user-visible behavior should be the same whether or not inference gives you a precise type or `Any`.